src/serializer/serializer.js

```diff
@@ -119,12 +120,45 @@ export class Serializer {
     if (timingStats !== undefined) timingStats.deepTraversalTime = Date.now();
     let residualHeapVisitor = new ResidualHeapVisitor(this.realm, this.logger, this.modules);
     residualHeapVisitor.visitRoots();
+    this.realm.wrapInGlobalEnv(() => this.realm.evaluateForEffects(() => {
```

Comment:
We aren't evaluating functions again, we're taking their effects from `writeEffects` and applying them in an evaluateForEffects context so that the changes are reversed afterwards. Maybe I should make a helper function in the realm that's more clearly named.